phantom-drift/introspection.proto

```diff
@@ -0,0 +1,351 @@
+    // See docs for SwarmCounterGroup for inner fields.
+    message TaggedSwarmCounterGroup {
+        map<string, TimeWindowCounter> incoming_opened = 1;
+        map<string, TimeWindowCounter> outgoing_opened = 2;
+        map<string, TimeWindowCounter> closed = 3;
+        map<string, uint32> active = 4;
+    }
+
+    // connections statistics.
+    SwarmCounterGroup conn_stats = 1;
+    // stream statistics
+    SwarmCounterGroup stream_stats = 2;
+    // stream statistics per protocol.
+    TaggedSwarmCounterGroup stream_by_proto_stats = 3;
+
+    // connection listing.
+    repeated Connection conns = 4;
+    // streams listing.
+    repeated Stream streams = 5;
+
+    // inner subsystems
+    Dialer dialer = 6;
+
+    // We want to be frugal with the 1-15 field id range, as they are represented by 1 byte only.
+    reserved 7 to 15;
+}
+
+// Host is a top-level object conveying a cross-cutting view of the entire system, including all subsystems.
+// TODO: WIP right now we're only covering the Swarm subsystem.
+message Host {
+    // our peer id.
+    PeerId id = 1;
+    // the transports enabled in this host.
+    repeated Transport transports = 2;
+    // the protocols attached to this host.
+    repeated string protocols = 3;
+    // the swarm subsystem.
+    Swarm swarm = 4;
+}
+
+// EndpointPair is a pair of multiaddrs.
+message EndpointPair {
+    // the source multiaddr.
+    string src_multiaddr = 1;
+    // the destination multiaddr.
+    string dst_multiaddr = 2;
+}
+
+// The status of a connection or stream.
+enum Status {
+    ACTIVE  = 0;
+    CLOSED  = 1;
+    OPENING = 2;
+    CLOSING = 3;
+    ERROR   = 4;
+}
+
+// Our role in a connection or stream.
+enum Role {
+    INITIATOR = 0;
+    RESPONDER = 1;
+}
+
+// Connection reports metrics and state of a libp2p connection.
+message Connection {
+    // Timeline contains the timestamps of the well-known milestones of a connection.
+    message Timeline {
+        // the instant when a connection was opened on the wire.
+        google.protobuf.Timestamp open_ts = 1;
+        // the instant when the upgrade process (handshake, security, multiplexing) finished.
+        google.protobuf.Timestamp upgraded_ts = 2;
+        // the instant when this connection was terminated.
+        google.protobuf.Timestamp close_ts = 3;
+    }
+
+    // Attributes encapsulates the attributes of this connection.
+    message Attributes {
+        // the multiplexer being used.
+        string multiplexer  = 1;
+        // the encryption method being used.
+        string encryption   = 2;
+    }
+
+    // the id of this connection, not to be shown in user tooling,
+    // used for (cross)referencing connections (e.g. relay).
+    bytes id                = 1;
+    // the peer id of the other party.
+    PeerId peer_id          = 2;
+    // the status of this connection.
+    Status status           = 3;
+    // the transport managing this connection.
+    Transport transport     = 4;
```

Comment:
shouldn't this be just the transport name? I think the `Transport` type has a lot of redundant information

Reply:
I changed this to a transport_id, and added a bytes id to Transport. How does that sound?